src/puppetlabs/puppetdb/cli/benchmark.clj

```diff
@@ -472,27 +431,156 @@
-  [& args]
+(defn random-hosts
+  [n pdb-host catalogs reports facts]
+  (let [random-entity (fn [host entities]
+                        (some-> entities
+                                rand-nth
+                                (assoc "certname" host)))]
+    (for [i (range n)]
+      (let [host (str "host-" i)]
+        {:host host
+         :catalog (when-let [cat (random-entity host catalogs)]
+                    (update cat "resources"
+                            (partial map #(update % "tags"
+                                                  conj
+                                                  pdb-host))))
+         :report (random-entity host reports)
+         :factset (random-entity host facts)}))))
+
+(defn populate-queue
+  "Fills queue with host entries.  Stops if mq is closed."
+  [mq numhosts run-interval pdb-host catalogs reports facts]
+  (println-err (trs "Populating queue in the background"))
+  (doseq [host (random-hosts numhosts pdb-host catalogs reports facts)
+          :while (>!! mq (assoc host :lastrun (rand-lastrun run-interval)))]
+    true)
+  (println-err (trs "Finished populating queue")))
+
+(defn close-and-drain-channel [channel]
+  "Closes channel and then drains it.  On exception, prints it to
+  *err* and returns false, otherwise returns true.  If channel is
+  already closed, returns true."
+  (try
+    (close! channel)
+    (while (<!! channel) true)
+    true
+    (catch Exception ex
+      (println-err ex)
+      false)))
+
+(defn delete-dir-or-report [dir]
+  (try
+    (fs/delete-dir dir)
+    (catch Exception ex
+      (println-err ex))))
+
+(defn start-filling-queue [mq numhosts run-interval pdb-host
+                           catalogs reports facts
+                           tmpdir]
+  "Starts a daemon thread to fill queue and returns a function to
+  request that the thread halt.  Ensures that tmpdir is eventually
+  deleted.  Upon error, tmpdir is deleted here, and upon success, it
+  will have been registered for deletion at shutdown."
+  (try
+    (println-err (trs "Using scratch dir {0}" (pr-str (str tmpdir))))
+    (let [fill-started (promise)
+          fill-queue (doto (Thread.
+                            #(do
+                               (deliver fill-started true)
+                               (populate-queue mq numhosts run-interval
+                                               pdb-host
+                                               catalogs reports facts)))
+                       (.setDaemon true))
+          ;; Use delay as one-shot
+          stop (delay
+                @fill-started ;; So we know it's OK to join
+                (let [result (do (.join fill-queue 500)
+                                 (or (not (.isAlive fill-queue))
+                                     (do
+                                       (println-err (trs "Waiting up to 10s for shutdown"))
+                                       ;; Do we need to make this longer b/c CI mess?
```

Comment:
Hmm, I'm actually leaning the other way (shorter) for the default since (currently) benchmark will be hard to kill during this period (i.e. kill -9) -- I'm also now wondering if we might not need to worry about this one as much; i'll ping you in a bit.